src/vm/genericdict.cpp

```diff
@@ -109,183 +109,333 @@ DictionaryLayout::GetFirstDictionaryBucketSize(
 // Optimize the case of a token being !i (for class dictionaries) or !!i (for method dictionaries)
 // 
 /* static */
-BOOL 
-DictionaryLayout::FindTokenWorker(LoaderAllocator *                 pAllocator,
-                                  DWORD                             numGenericArgs,
-                                  DictionaryLayout *                pDictLayout,
-                                  CORINFO_RUNTIME_LOOKUP *          pResult,
-                                  SigBuilder *                      pSigBuilder,
-                                  BYTE *                            pSig,
-                                  DWORD                             cbSig,
-                                  int                               nFirstOffset,
-                                  DictionaryEntrySignatureSource    signatureSource,
-                                  WORD *                            pSlotOut)
+BOOL DictionaryLayout::FindTokenWorker(LoaderAllocator*                 pAllocator,
+                                       DWORD                            numGenericArgs,
+                                       DictionaryLayout*                pDictLayout,
+                                       SigBuilder*                      pSigBuilder,
+                                       BYTE*                            pSig,
+                                       DWORD                            cbSig,
+                                       int                              nFirstOffset,
+                                       DictionaryEntrySignatureSource   signatureSource,
+                                       CORINFO_RUNTIME_LOOKUP*          pResult,
+                                       WORD*                            pSlotOut,
+                                       DWORD                            scanFromSlot /* = 0 */,
+                                       BOOL                             useEmptySlotIfFound /* = FALSE */)
+
 {
     CONTRACTL
     {
         STANDARD_VM_CHECK;
         PRECONDITION(numGenericArgs > 0);
+        PRECONDITION(scanFromSlot >= 0 && scanFromSlot <= pDictLayout->m_numSlots);
         PRECONDITION(CheckPointer(pDictLayout));
-        PRECONDITION(CheckPointer(pSlotOut));
+        PRECONDITION(CheckPointer(pResult) && CheckPointer(pSlotOut));
         PRECONDITION(CheckPointer(pSig));
         PRECONDITION((pSigBuilder == NULL && cbSig == -1) || (CheckPointer(pSigBuilder) && cbSig > 0));
     }
     CONTRACTL_END
 
-#ifndef FEATURE_NATIVE_IMAGE_GENERATION
-    // If the tiered compilation is on, save the fast dictionary slots for the hot Tier1 code
-    if (g_pConfig->TieredCompilation() && signatureSource == FromReadyToRunImage)
```

Comment:
It is really worth it to allocate the actual dictionary slots with tiered compilation? They will just waste memory in steady state.

Reply:
I'm not seeing a significant increase in memory usage with these changes when I tested using a WPF app.

Reply:
WPF apps are so big that anything we do in the runtime is going be a drop in the bucket. It is not a good reason to just doing wasteful things in the runtime.

Reply:
I can collect some numbers for the webapi sample and check the impact of memory.

Reply:
I have been thinking about it some more. I think it is probably fine to always allocate the actual dictionary slots. Feel free to ignore this discussion.